lib-rpc/ocamlformat_rpc_lib.mli

```diff
@@ -9,70 +9,78 @@
 (*                                                                        *)
 (**************************************************************************)
 
-module type Command_S = sig
-  type t
+module Make (IO : IO.S) : sig
+  module type Command_S = sig
+    type t
 
-  val read_input : Stdlib.in_channel -> t
+    val read_input : IO.ic -> t IO.t
 
-  val to_sexp : t -> Sexplib0.Sexp.t
+    val to_sexp : t -> Csexp.t
 
-  val output : Stdlib.out_channel -> t -> unit
-end
+    val output : IO.oc -> t -> unit IO.t
+  end
 
-module type Client_S = sig
-  type t
+  module type V = sig
+    module Command : Command_S
 
-  type cmd
+    module Client : sig
+      type t
 
-  val pid : t -> int
+      type cmd
 
-  val mk : pid:int -> in_channel -> out_channel -> t
+      val pid : t -> int
 
-  val query : cmd -> t -> cmd
+      val mk : pid:int -> IO.ic -> IO.oc -> t
 
-  val halt : t -> (unit, [> `Msg of string]) result
+      val query : cmd -> t -> cmd IO.t
 
-  val config :
-    (string * string) list -> t -> (unit, [> `Msg of string]) result
+      val halt : t -> (unit, [> `Msg of string]) result IO.t
+      (** The caller must close the input and output channels after calling
+          [halt]. *)
 
-  val format : string -> t -> (string, [> `Msg of string]) result
-end
+      val config :
+        (string * string) list -> t -> (unit, [> `Msg of string]) result IO.t
 
-module type V = sig
-  module Command : Command_S
+      val format : string -> t -> (string, [> `Msg of string]) result IO.t
+    end
+    with type cmd = Command.t
+  end
 
-  module Client : Client_S with type cmd = Command.t
-end
+  (** Version used to set the protocol version *)
+  module Init :
+    Command_S with type t = [`Halt | `Unknown | `Version of string]
```

Comment:
This `Init` module isn't useful for users of the library. The server uses it but I think it should be defined there. (same for `Command_S` and `V1.Command`.

To be sure the server and the lib are using the same types, there can be a module dedicated to that, which doesn't pretend to be part of the API:
```ocaml
  (** Types used for communicating between the client and the server. *)
  module Commands : sig
    module V1 : sig
      type init = [`Halt | `Unknown | `Version of string]
      type t =
        [ `Halt
        | `Unknown
        | `Error of string
        | `Config of (string * string) list
        | `Format of string ]
    end
  end
```

Reply:
`Init` is useful for the server `bin/ocamlformat-rpc/main.ml` that is also a server of the library unfortunately. Each type of command also needs the read/output/to_sexp functions, that are parameterized by IO, so I don't find a better architecture than the current one for now.